engine/gamesys/src/gamesys/scripts/script_resource.cpp

```diff
@@ -460,6 +461,241 @@ static int GraphicsTextureTypeToImageType(int texturetype)
     return -1;
 }
 
+static dmResource::Result GetTextureDataCb(const dmResource::GetResourceDataCallbackParams* params)
+{
+    *params->m_Message = (dmGraphics::TextureImage*) params->m_UserData;
+    return dmResource::RESULT_OK;
+}
+
+static dmGraphics::TextureImage MakeTextureImage(uint16_t width, uint16_t height, uint8_t max_mipmaps, uint8_t bitspp,
+    dmGraphics::TextureImage::Type type, dmGraphics::TextureImage::TextureFormat format)
+{
+    uint32_t* mip_map_sizes   = new uint32_t[max_mipmaps];
+    uint32_t* mip_map_offsets = new uint32_t[max_mipmaps];
+    uint8_t layer_count       = type == dmGraphics::TextureImage::TYPE_CUBEMAP ? 6 : 1;
+
+    uint32_t data_size = 0;
+    uint16_t mm_width  = width;
+    uint16_t mm_height = height;
+    for (uint32_t i = 0; i < max_mipmaps; ++i)
+    {
+        mip_map_sizes[i]   = dmMath::Max(mm_width, mm_height);
+        mip_map_offsets[i] = (data_size / 8);
+        data_size += mm_width * mm_height * bitspp * layer_count;
+        mm_width  /= 2;
+        mm_height /= 2;
+    }
+    assert(data_size > 0);
+
+    data_size                *= layer_count;
+    uint32_t image_data_size  = data_size / 8; // bits -> bytes for compression formats
+    uint8_t* image_data       = new uint8_t[image_data_size];
+
+    dmGraphics::TextureImage::Image* image = new dmGraphics::TextureImage::Image();
+    dmGraphics::TextureImage texture_image = {};
+    texture_image.m_Alternatives.m_Data    = image;
+    texture_image.m_Alternatives.m_Count   = 1;
+    texture_image.m_Type                   = type;
+
+    image->m_Width                = width;
+    image->m_Height               = height;
+    image->m_OriginalWidth        = width;
+    image->m_OriginalHeight       = height;
+    image->m_Format               = format;
+    image->m_CompressionType      = dmGraphics::TextureImage::COMPRESSION_TYPE_DEFAULT;
+    image->m_CompressionFlags     = 0;
+    image->m_Data.m_Data          = image_data;
+    image->m_Data.m_Count         = image_data_size;
+
+    image->m_MipMapOffset.m_Data  = mip_map_offsets;
+    image->m_MipMapOffset.m_Count = max_mipmaps;
+    image->m_MipMapSize.m_Data    = mip_map_sizes;
+    image->m_MipMapSize.m_Count   = max_mipmaps;
+
+    return texture_image;
+}
```

Comment:
Looking into it a bit further, perhaps you can try the way the engine packs other ddf messages. E.g. from `gameobject/script.h`:

```
    template <typename TDDFType>
    Result PostDDF(const TDDFType* message, const dmMessage::URL* sender, const dmMessage::URL* receiver, int function_ref, bool unref_function_after_call)
    {
        dmArray<uint8_t> packed_message;
        packed_message.SetCapacity(sizeof(TDDFType)); // A good estimate for most messages
        dmDDF::SaveMessageToArray(message, message->m_DDFDescriptor, packed_message);
        return PostScriptMessage(message->m_DDFDescriptor, packed_message.Begin(), packed_message.Size(), sender, receiver, function_ref, unref_function_after_call);
    }
```

Ofc, that is the generic way of doing it.
And I suspect that it will make a full copy of the texture data.
But perhaps it's a good first start, we can make a more optimized packer later.
And I think it will remove a lot of the extra "fluff" introduced by the mix between messages and raw pointers.